### PR TITLE
Add cuffdiff helper function

### PR DESCRIPTION
**Pull request by jvrakor**
@dblenkus @jkokosar

Add cuffdiff helper function.

### CODE REVIEW

File: resdk/analysis/differential_expressions.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        if set(sample_ids_rel).issubset(sample_ids_res):
+            # maybe add a flag here and report error at the end corresponding to
+            # no relations containing selected samples.
+            for sample, position in zip(relation.samples, relation.positions):
+                if position == 'case':
+                    case.append(get_data_id(sample.get_cuffquant()))
+                elif position == 'control':
+                    control.append(get_data_id(sample.get_cuffquant()))
+                else:
+                    raise LookupError(
```
Comment (dblenkus) [inline resdk/analysis/differential_expressions.py]: I think that `ValueError` would be more appropriate here.

https://docs.python.org/3/library/exceptions.html

File: resdk/analysis/differential_expressions.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        inputs['threads'] = threads
+
+    relation_filter = {}
+    collection_id = get_resource_collection(resource)
+    if collection_id:
+        relation_filter['collection'] = collection_id
+
+    samples = get_samples(resource)
+
+    input_objects = [annotation]
```
Comment (dblenkus) [inline resdk/analysis/differential_expressions.py]: Also add `genome` to `input_objects`.

File: resdk/analysis/differential_expressions.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    input_objects.extend(samples)
+    resolwe = get_resolwe(*input_objects)
+
+    relations = resolwe.relation.filter(
```
Comment (dblenkus) [inline resdk/analysis/differential_expressions.py]: This query must be further filtered. Imagine the case where `relation_filter` is empty (i.e. `cuffdiff` is called on a list of samples), this will pull **all** `case-control` relations that exists on the server.

File: resdk/tests/functional/analysis/e2e_differential_expressions.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    def test_cuffdiff(self):
+        collection = self.res.collection.create(name='Test collection')
+
+        # pylint: disable=unbalanced-tuple-unpacking
+        cuffquant_1, cuffquant_2, cuffquant_3, cuffquant_4 = self.get_cuffquants(4, collection)
+        # pylint: enable=unbalanced-tuple-unpacking
+        gff = self.get_gtf(1)
+
+        relation = collection.create_compare_relation(
+            samples=[cuffquant_1.sample, cuffquant_2.sample, cuffquant_3.sample,
```
Comment (dblenkus) [inline resdk/tests/functional/analysis/e2e_differential_expressions.py]: Write this as:

    samples=[
        cuffquant_1.sample,
        cuffquant_2.sample,
        cuffquant_3.sample,
        cuffquant_4.sample,
    ],

File: resdk/tests/functional/analysis/e2e_differential_expressions.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        )
+
+        # Run cuffdiff on a collection
+        cuffdiff = collection.run_cuffdiff(annotation=gff)
+        self.assertEqual(cuffdiff.input['case'], [cuffquant_3.id, cuffquant_4.id])
+        self.assertEqual(cuffdiff.input['control'], [cuffquant_1.id, cuffquant_2.id])
+        self.assertEqual(cuffdiff.input['annotation'], gff.id)
+
+        # Run cuffdiff on a relation
+        cuffdiff = differential_expressions.cuffdiff(relation, annotation=gff)
```
Comment (dblenkus) [inline resdk/tests/functional/analysis/e2e_differential_expressions.py]: You can call this as `relation.run_cuffdiff(annotation=gff)`

File: resdk/tests/functional/analysis/e2e_differential_expressions.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+class TestExpressions(BaseResdkFunctionalTest):
+
+    def test_cuffdiff(self):
+        collection = self.res.collection.create(name='Test collection')
+
+        # pylint: disable=unbalanced-tuple-unpacking
+        cuffquant_1, cuffquant_2, cuffquant_3, cuffquant_4 = self.get_cuffquants(4, collection)
+        # pylint: enable=unbalanced-tuple-unpacking
+        gff = self.get_gtf(1)
```
Comment (dblenkus) [inline resdk/tests/functional/analysis/e2e_differential_expressions.py]: This `1` is id of collection to which gtf objects will be added. You probably don't want to do this.

Use either `self.get_gtf(collection.id)` or `self.get_gtf()`.

File: resdk/tests/functional/analysis/e2e_differential_expressions.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+# pylint: disable=missing-docstring
```
Comment (dblenkus) [inline resdk/tests/functional/analysis/e2e_differential_expressions.py]: Write some more tests that covers edge cases. I.e. what happens if some samples from the collections are not in any relation,...

File: resdk/analysis/differential_expressions.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        label='case-control',  # this might not be needed
+        **relation_filter
+    )
+
+    for relation in relations:
+        control = []
+        case = []
+
+        sample_ids_rel = [sample.id for sample in get_samples(relation)]
+        sample_ids_res = [sample.id for sample in samples]
```
Comment (dblenkus) [inline resdk/analysis/differential_expressions.py]: You can move this line out of for as it doesn't affect it.

File: resdk/analysis/differential_expressions.py (hunk 3 above)
Comment (jvrakor) [inline resdk/analysis/differential_expressions.py]: This query will pull all case-control relations that are in any way related to the samples (i.e. any sample from the resource is in this relation). This is further connected to skipping unsuitable relations mentioned later. But most important: all `case-control` relations from the server will NOT be pulled.

Comment (jvrakor) [inline resdk/analysis/differential_expressions.py]: This is connected to skipping unsuitable relations mentioned later. That is also the fail-safe. We can of course decide on a different implementation.

File: resdk/analysis/differential_expressions.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    )
+
+    for relation in relations:
+        control = []
+        case = []
+
+        sample_ids_rel = [sample.id for sample in get_samples(relation)]
+
+        if not set(sample_ids_rel).issubset(sample_ids_res):  # optional, not yet decided
+            continue
```
Comment (dblenkus) [inline resdk/analysis/differential_expressions.py]: I think it would be better to skip (in next `for`) just samples that are not in `samples`.

File: resdk/analysis/differential_expressions.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        inputs['library_normalization'] = library_normalization
+
+    if dispersion_method is not None:
+        inputs['dispersion_method'] = dispersion_method
+
+    if threads is not None:
+        inputs['threads'] = threads
+
+    samples = get_samples(resource)
+    sample_ids_res = [sample.id for sample in samples]
```
Comment (dblenkus) [inline resdk/analysis/differential_expressions.py]: Rename this to `sample_ids`.

File: resdk/analysis/differential_expressions.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        **relation_filter
+    )
+
+    cuffdiff_objects = []
+    for relation in relations:
+        control = []
+        case = []
+        for sample, position in zip(relation.samples, relation.positions):
+            if sample.id not in sample_ids_res:
+                continue
```
Comment (dblenkus) [inline resdk/analysis/differential_expressions.py]: Add a new line after `continue`.

File: resdk/tests/functional/analysis/e2e_differential_expressions.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from resdk.analysis import differential_expressions
+from resdk.tests.functional.base import BaseResdkFunctionalTest
+
+
+class TestExpressions(BaseResdkFunctionalTest):
+
+    def test_cuffdiff(self):
+        collection = self.res.collection.create(name='Test collection')
```
Comment (dblenkus) [inline resdk/tests/functional/analysis/e2e_differential_expressions.py]: Rename these to `collection_1` and `collection_2`.

**Comment by mstajdohar** [comment]
Write tests to improve coverage.

**Comment by dblenkus** [comment]
There are nice e2e tests for added functionality. Problem is that they are run on Jenkins, but coverage report is not pushed to codecov.

@tjanez can you fix this?
